### ISSUE TITLE
jailbreakd memory issues

### ISSUE THREAD

**Issue by jakeajames** [pwn20wndstuff/Undecimus#12]
A while ago I discovered a way to trigger Electra's freezing by overloading WebKit till it crashed (Twitter zoom-in & out bug; see: https://github.com/coolstar/electra1131/issues/14)

After testing this on unc0ver the first time I observed the same behavior that happens on stock iOS, WebKit crashes and device continues to behave normally. However I got some mentions on Twitter from people who tried it and either got a freeze or respring loop (both happened with Electra). I gave it a few more tries, the second try it indeed freezed and the third it did not.

Then I decided to add some logging on the jailbreakd and got some interesting results. Immediately after the webkit crash, random daemons also crash, jailbreakd gets flooded with requests until it gets restarted due to reaching memory limits. The issue is definitely different and not as bad as Electra but still exists.

Here are the full logs:
https://pastebin.com/raw/xGX5rhs0

**Comment by jakeajames** [comment]
@Co2333 it is indeed a bug, as webkit crashes in stock mode as well, however in stock system recovers from the crash without issues, here jailbreakd gets overloaded and restarts, restarting jailbreakd is not always a good idea; this can also leave certain processes frozen (suspended) due to crashing before sending SIGCONT

**Comment by KMamedoff** [comment]
It is a bug that no one takes seriously for some reason I don't understand.

**Comment by parrotgeek1** [comment]
@jakeajames this issue is caused by iOS killing jailbreakd when it runs out of ram, not jailbreakd itself. Coolstar fixed it but explicitly refuses (he told me this) to release the fixed build because he hates the community.

**Comment by sbingner** [comment]
I fixed it to the extent it is fixed by making launchd pull the jbd port again if it’s invalid... but it probably needs a little more testing.  Pspawn in launchd used to only pull the jbd port when it was first injected

**Comment by parrotgeek1** [comment]
Just curious; why is jailbreakd not just running inside launchd?

**Comment by jakeajames** [comment]
@parrotgeek1 Not because RAM gets full but jailbreakd's memory limit is reached, a lot of daemons crash at once and jailbreakd gets a lot of requests, which makes it get killed by launchd. The only place I can think of where a lot of memory is being used is PROC_PIDPATHINFO_MAXSIZE (4KB), but still I don't see why jailbreakd would crash from a lot of requests, unless stack allocates that much memory for every single request?

Nevertheless, jailbreakd crashing is not a good thing for many reasons, kernel execution never gets terminated and memory used by it on the kernel does not get freed, (and we allocate that memory each time we load jailbreakd); also if jailbreakd happens to crash while processing a request it can leave that process stuck in a frozen state, since nothing sends the CONT signal to it.

**Comment by KMamedoff** [comment]
> @jakeajames this issue is caused by iOS killing jailbreakd when it runs out of ram, not jailbreakd itself. Coolstar fixed it but explicitly refuses (he told me this) to release the fixed build because he hates the community.

If he is aware of this issue then why does he always share his uptime with Electra? It does not make sense to me.

**Comment by parrotgeek1** [comment]
@kmamedoff I have no idea. I've tried to convince him privately to release the fixed version (he literally even sent me a screenshot of the ipa) but he won't because "the community hates him"

**Comment by KMamedoff** [comment]
> @KMamedoff I have no idea. I've tried to convince him privately to release the fixed version (he literally even sent me a screenshot of the ipa) but he won't because "the community hates him"

No wonder why community hates him after releasing a buggy jailbreak and never fixing it.

**Comment by Samgisaninja** [comment]
Alright folks, let’s try to keep this on-topic please. If you’d like to discuss how other people have handled (or failed to handle) this issue, take that to reddit.

**Comment by Co2333** [comment]
Okay, so one probably solution is to send -CONT signal to process that looks like being freezed or just kill it if it is no much matter retaliated to system.
And a better solution is to write a WebKit guard which kills WebKit when sending too much request to jailbreakd. 

I will continue doing research on it by simulating the event on a sandboxed app and trying to -CONT each process.

**Comment by jakeajames** [comment]
@Co2333 the answer that makes the most sense is that launchd is killing jailbreakd due to memory pressure, nothing specific to WebKit. A way to solve would be either properly restart jailbreakd and clean up memory used by previous session, or make jailbreakd "unkillable" or "important" somehow.

**Comment by Co2333** [comment]
@jakeajames It is not killing jailbreakd anyhow on my device.

![image](https://user-images.githubusercontent.com/25259084/47373066-a5c56280-d71d-11e8-8786-2829fc6a2c76.png)

This time jailbreakd's PID is 6122 and it is not changed anyway after a jailbreak. Unc0ver one.
I triggered this issue using twitter's method and it went into reloading for the first time. The second time is a respring loop. I was so busy at debugging the info showing on my computer while my iPad doing its job. I noticed that charge sound play twice. When I look back to my iPad, it is alive.
The kernel is not killing jailbreakd but some process called "webbookmarksd" . This process is the one which being killed most of the time. And others like "com.apple.safari.ImageDecoder" is also being killed. Any relation to UI render maybe? Only a thought.

----Basically, we went back to were we came from.

This might be something like a Dispatch Queue obstruct. But luckily, those queues all have a timeout. If you wait for enough time, no reboot. I'm trying to free these queues manually to see if it works.

----This is my condition.

**Comment by Co2333** [comment]
Trying to run killall xpcproxy. It works for me.

**Comment by Samgisaninja** [comment]
2bb32ff and later removes jailbreakd entirely if Cydia Substrate is installed, so even though this issue still exists, it can be fixed by removing substitute and using substrate for code injection. Not exactly a fix for the issue, "workaround" (?)

**Comment by KMamedoff** [comment]
What causes the issue? Has it been found?

**Comment by jakeajames** [comment]
@Samgisaninja substitute got nothing more than substrate, while substrate does have a lot of advantages so I don't think it'll be a problem. Even if you pushed everyone there, that wouldn't be bad, afterall all ios 10 and below jailbreaks used substrate only. ¯\_(ツ)_/¯

**Comment by resistor4u** [comment]
On iOS 11.3.1 on iphone 6s, when running any apt-related command (e.g., `apt-get update` or `apt-cache search` or `aptitude install` or `apt-get autoremove --purge`) it fails with `Segmentation fault: 11.. 89%` (usually between 87-93%). No issues with cydia. Do the seg-faults have something to do with the memory issue?

Here's my `df -h`
```
Filesystem       Size   Used  Avail Capacity iused               ifree %iused  Mounted on
/dev/disk0s1s1   15Gi  3.5Gi  980Mi    79%   68549 9223372036854707258    0%   /
devfs            50Ki   50Ki    0Bi   100%     172                   0  100%   /dev
/dev/disk0s1s2   15Gi   10Gi  980Mi    92%   70425 9223372036854705382    0%   /private/var
/dev/disk0s1s3   10Mi  1.0Mi  9.0Mi    11%       3 9223372036854775804    0%   /private/var/wireless/baseband_data
```

**Comment by Samgisaninja** [comment]
Uh, no.

1. “Memory” refers to RAM, not your SSD
2. apt has nothing to do with any of this

**Comment by resistor4u** [comment]
i'm aware ram and device storage are different things, but TIL the ios doesn't use swapfiles. [a log ](https://pastebin.com/wg8qv7Bs) of the crashes suggests that memory is involved, insofar as the addresses don't exist or access isn't provided (right?). either way, it doesn't seem it's involved with ram getting full.

**Comment by KMamedoff** [comment]
That bug is still there even with Cydia Substrate 0.9.7000. Device just freezes and SSH stops working.

**Comment by Ultra03** [comment]
@KMamedoff Likely not jailbreakd because Substrate doesn't use it, but yes - it is confirmed that Substrate is freezing A7 and A8 devices for some reason.

**Comment by KMamedoff** [comment]
> @KMamedoff Likely not jailbreakd because Substrate doesn't use it, but yes - it is confirmed that Substrate is freezing A7 and A8 devices for some reason.

I am glad that it is confirmed. I was uninstalling tweaks one by one with no luck. Do you know who confirmed it or is saurik working on it?

**Comment by Ultra03** [comment]
>I am glad that it is confirmed. I was uninstalling tweaks one by one with no luck. Do you know who >confirmed it or is saurik working on it?

I talked to him on Discord about it:
![image01](https://user-images.githubusercontent.com/20672260/50188102-4034c300-02d5-11e9-8bb7-b28f493c982a.png)

**Comment by KMamedoff** [comment]
> > I am glad that it is confirmed. I was uninstalling tweaks one by one with no luck. Do you know who >confirmed it or is saurik working on it?
> 
> [I talked to him on Discord about it](https://user-images.githubusercontent.com/20672260/50188102-4034c300-02d5-11e9-8bb7-b28f493c982a.png)

I hope it gets fixed soon. Thank you for screenshot.

**Comment by pwn20wndstuff** [comment]
Fixed in Substrate 0.9.7010! 🎉 - Thanks to @saurik

**Comment by KMamedoff** [comment]
> Fixed in Substrate 0.9.7010! 🎉 - Thanks to @saurik

Does it fix freezing issues on A7/A8 devices?

**Comment by Diatrus** [comment]
This issue should probably be closed? @jakeajames
